File: El Servador/god_kaiser_server/src/services/ledger_ec_units.py
"""
Ledger EC unit adapter (AUT-1350 / U1).

Ledger DB columns (``ec_*_ms_cm``, ``nutrient_solution_batches``) store **mS/cm**.
Operational fertigation (Plan / Assist / Logic / FE targets) uses **µS/cm** SSOT.

ALL ×1000 / ÷1000 between these worlds MUST go through this module — one boundary.
Do not inline scale factors in Assist, dose calculators, or FE.
"""

from __future__ import annotations

from typing import Optional

# Exact factor: 1 mS/cm = 1000 µS/cm
US_PER_MS = 1000.0


def ledger_ms_cm_to_us_cm(ms_cm: float) -> float:
    """Read boundary: Ledger mS/cm → operational µS/cm."""
    return float(ms_cm) * US_PER_MS


def us_cm_to_ledger_ms_cm(us_cm: float) -> float:
    """Write boundary: operational µS/cm → Ledger mS/cm."""
    return float(us_cm) / US_PER_MS


def optional_ledger_ms_cm_to_us_cm(ms_cm: Optional[float]) -> Optional[float]:
    """Nullable read helper for Assist / composition ledger reads."""
    if ms_cm is None:
        return None
    return ledger_ms_cm_to_us_cm(ms_cm)


def optional_us_cm_to_ledger_ms_cm(us_cm: Optional[float]) -> Optional[float]:
    """Nullable write helper when persisting operational EC into Ledger."""
    if us_cm is None:
        return None
    return us_cm_to_ledger_ms_cm(us_cm)

File: El Frontend/src/utils/ledgerEcUnits.ts
/**
 * Ledger EC unit adapter (AUT-1350 / AUT-1358) — FE mirror of
 * `El Servador/.../services/ledger_ec_units.py`.
 *
 * Ledger API/DB fields (`ec_*_ms_cm`, batch `ec_measured_after`) store **mS/cm**.
 * Operational FE (Plan / Ist-Soll / Batch-Modal input) uses **µS/cm** SSOT.
 *
 * ALL ×1000 / ÷1000 at this boundary MUST go through these helpers.
 */

/** Exact factor: 1 mS/cm = 1000 µS/cm (same as server US_PER_MS). */
export const US_PER_MS = 1000

/** Read boundary: Ledger mS/cm → operational µS/cm. */
export function ledgerMsCmToUsCm(msCm: number): number {
  return Number(msCm) * US_PER_MS
}

/**
 * Write boundary: operational µS/cm → Ledger mS/cm.
 * Mirrors server `us_cm_to_ledger_ms_cm` / `TankService.to_ledger_ec_ms_cm`.
 */
export function usCmToLedgerMsCm(usCm: number): number {
  return Number(usCm) / US_PER_MS
}

/** Nullable write helper before POST …/batches. */
export function optionalUsCmToLedgerMsCm(usCm: number | null | undefined): number | null {
  if (usCm == null) return null
  return usCmToLedgerMsCm(usCm)
}
